src/datasets/utils/file_utils.py

```diff
@@ -620,16 +507,16 @@ def _resumable_file_manager():
 
             # GET file object
             if url.startswith("ftp://"):
-                ftp_get(url, temp_file, proxies=proxies, resume_size=resume_size, headers=headers, cookies=cookies)
+                FtpClient.get(url, callback=lambda r: shutil.copyfileobj(r, temp_file))
             else:
-                http_get(
+                HttpClient.get(
```

Comment:
I feel like the callback logic is not explicit, it makes the code harder to understand imho. This is not javascript.

I would move here all this logic from the HttpClient.get:
```python
            if response.status_code == 416:  # Range not satisfiable
                return
            if callback:
                content_length = response.headers.get("Content-Length")
                total = resume_size + int(content_length) if content_length is not None else None
                not_verbose = bool(logger.getEffectiveLevel() > WARNING)
                progress = tqdm(
                    unit="B",
                    unit_scale=True,
                    total=total,
                    initial=resume_size,
                    desc="Downloading",
                    disable=not_verbose,
                )
                for chunk in response.iter_content(chunk_size=1024):
                    if chunk:  # filter out keep-alive new chunks
                        progress.update(len(chunk))
                        callback(chunk)
```
So that it's the _resumable_file_manager that is responsible for the progress bar and the status codes handling.

I'd be happy to discuss that though.

I think a HttpClient should be quite bare-bone and not do additional magic.
Simply taking care of the header, handling retries and returning the response is enough.

Another possible option would be to give pass as a parameter `local_filename` where to write the data. This option would keep the for loop inside the HttpClient.get and would remove the callback logic. This also has the advantage that we can do the same thing for the FtpClient to have consistent behaviors. Let me know what option sounds best for you.

Reply:
Hi @lhoestq . Thanks a lot for your critical review: that is exactly the way we will further improve Hugginface Datasets!!! :)

I agree with you: I do not like this "callback" logic either. It is not Pythonic and it is confusing (as you said right, adding hidden logic under the hood). 

I was somehow induced to implement it because of the original call hierarchy:
- temp local file opening
  - remote http/ftp call
     - use again temp local file logic but from within the remote caller scope
As you see, "calback" was a way to be able to pass logic from the outer scope to the inner one...

However, I was not happy with the resulting implementation. "There must be a better way...". And thinking about this during the weekend, I had a new idea: let's change the call hierarchy!!!

I am refactoring all this and once done, let me know your opinion... :)

Reply:
Changing the hierarchy order sounds good ! :)